runtime/src/main/scala/grpcgateway/util/UrlTemplate.scala

```diff
@@ -0,0 +1,78 @@
+package grpcgateway.util
+
+import io.netty.handler.codec.http.QueryStringDecoder
+
+import scala.collection.mutable
+
+/** A means of parsing request URLs with support for "URL parameter templates" configured in a protobuf descriptor */
+trait UrlTemplate {
+  /**
+    * Match incoming URI against this URL template generated from a protobuf RESTful service descriptor
+    *
+    * Netty's FullHttpRequest.uri() happens to return a path so we currently assume the "protocol/host" prefix
+    * is stripped before this method is called.
+    *
+    * @return URL properties extracted from the URI if it matched this template
+    */
+  def matchUri(uri: String) : Option[RestfulUrl]
+}
+
+/** Fast path for URL patterns with no templates */
+private final class PlainUrlTemplate(path: String) extends UrlTemplate {
+  override def matchUri(uri: String): Option[RestfulUrl] = {
+    //println(s"Matching \'$uri\' to $path")
+
+    val decoder = new QueryStringDecoder(uri)
+    if (decoder.path() == path) {
+      Some(new PlainRestfulUrl(decoder.parameters()))
+    } else {
+      None
+    }
+  }
+}
+
+/** Remember a sequence of matchers to apply (in the same order) to an incoming URI. While matching, optimistically
+  * collect values at positions corresponding to names slots in the original URL template */
+private final class MatchingUrlTemplate(matchers: Seq[PathMatcher]) extends UrlTemplate {
+  override def matchUri(uri: String): Option[RestfulUrl] = {
+    val decoder = new QueryStringDecoder(uri)
+    val path = decoder.path()
+
+    //println(s"Matching \'$path\' with ${matchers.mkString}")
+
+    var pathIndex = 0
+    var matcherIndex = 0
+
+    val templateParams = mutable.Map[String, String]()
+    while (pathIndex < path.length) {
+      val from = pathIndex
+
+      val matcher = matchers(matcherIndex)
```

Comment:
there is a bug here. in cases when matchers.length < path.length this line will thrown IndexOutOfBoundsException.